pyadjoint/tape.py

```diff
@@ -82,17 +100,23 @@ class stop_annotating(object):
     modified variables at the end of the context manager. """
 
     def __init__(self, modifies=None):
-        global _annotation_enabled
         self.modifies = modifies
-        self._orig_annotation_enabled = _annotation_enabled
+        # the `no_annotations` context manager could be nested,
+        # so we need a stack to keep track of the original states.
```

Comment:
Is this really necessary? I helped implement this originally and I thought that using `_orig_annotation_enabled` was enough to simulate a stack.

Reply:
Previously, the wrapper that `no_annotations` returned would instantiate a new instance of `stop_annotating` for every call of the decorated function. The "stack" was just lots of different instances of `stop_annotating`.

Now, because `no_annotations` is a single instance of `stop_annotating`, if you nest two functions decorated with `no_annotation`, then they both enter/exit the _same_ instance of `stop_annotating`. Hence we need an explicit stack.

Reply:
OK that makes some sense.